src/Build/Graph/ProjectGraph.cs

```diff
@@ -219,47 +218,100 @@ public ProjectGraph(string entryProjectFile)
             return graphNode;
         }
 
+        private enum NodeState
+        {
+            // the project has been evaluated and it's project references are being processed
+            InProcess,
+            // all project references of this project have been processed
+            Processed
+        }
+
         /// <summary>
         /// Load a graph with root node at entryProjectFile
-        /// Maintain a queue of projects to be processed- each queue item is a key value pair of the project to be evaluated and its parent
-        /// Once the project has been evaluated, add a project reference to this evaluated target from the parent node
+        /// Maintain the state of each node (InProcess and Processed) to detect cycles
+        /// returns false if loading the graph is not successful
         /// </summary>
-        private void LoadGraph(string entryProjectFile, ProjectCollection projectCollection, PropertyDictionary<ProjectPropertyInstance> globalProperties)
+        private (bool success, List<string> projectsInCycle) LoadGraph(ConfigurationMetadata projectToEvaluate,
```

Comment:
Since this is now a recursive algorithm, I think "ProcessNode" might be a better name for this method.